### ISSUE TITLE
Investigate exposing Arrow StringBuilder to numba

### ISSUE THREAD

**Issue by chmp** [xhochy/fletcher#2]
The numba implementation should be much slower, than arrow's StringBuilder.

Question: what can numba inteface with? The docs mention [cffi](http://numba.pydata.org/numba-doc/0.37.0/reference/pysupported.html#cffi).

**Comment by alendit** [comment]
I've looked into it recently: the fletcher's `NumbaStringArrayBuilder` is around 10 times slower than Arrow's `StringBuilder` for code like this for the current version of numba:

```
@numba.njit(nogil=True)
def build_string(chars):
   sb = fr._numba_compat.NumbaStringArrayBuilder(10 ** 6, 10 ** 7)
   for str_idx in range(0, 10 ** 6):
      for c_idx in range(0, 10):
         sb.put_byte(chars[str_idx * 10 + c_idx])
      sb.finish_string()
   sb.finish()
```

This comes from the issues with inlining and optimization of numba's `jitclasses`. 

Good news is that I'm working on a set of patches which will bring down the runtime of the code above from 400ms to 40ms, i.e. on par wit this C++ code. The relevant issues are https://github.com/numba/numba/issues/2166, https://github.com/numba/numba/issues/3305. This PR fixes the inlining https://github.com/numba/numba/pull/3531. This PR  https://github.com/numba/llvmlite/pull/429 to `llvmlite` will allow linking custom optimization passes, and finally there is local code which optimizes ref counting across the CFG.

TL;DR It can take a while, but there is not reason why numba wouldn't be as fast as native C++ for this task.

**Comment by xhochy** [comment]
@alendit That sounds promising. Thank you for taking a look into this!

**Comment by xhochy** [comment]
This project has been archived as development has ceased around 2021.
With the support of [Apache Arrow-backed extension arrays in `pandas`](https://github.com/pandas-dev/pandas/pull/35259), the major goal of this project has been fulfilled.
